### PR TITLE
Ex 12 register local

### PR DESCRIPTION
**Pull request by tomaszferens**
image

image

image

image

Demo: https://share.getcloudapp.com/d5ub0xNW

### CODE REVIEW

File: backend/src/auth/jwt.strategy.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+export class JwtStrategy extends PassportStrategy(Strategy) {
+    constructor() {
+        super({
+            jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
+            ignoreExpiration: false,
+            secretOrKey: jwtConstants.secret,
+        })
+    }
+
+    async validate(payload: any) {
```
Comment (marckraw) [inline backend/src/auth/jwt.strategy.ts]: should we add type for payload ? (if multiple possible, than we should have `SomeTypes || SomeTypes2` i guess)

File: backend/src/auth/local.strategy.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+import { AuthService } from './auth.service'
+
+@Injectable()
+export class LocalStrategy extends PassportStrategy(Strategy) {
+    constructor(private readonly authService: AuthService) {
+        super()
+    }
+
+    async validate(username: string, password: string): Promise<any> {
```
Comment (marckraw) [inline backend/src/auth/local.strategy.ts]: return type as with others :) Let's try to add them!

File: frontend/src/Landing.js [hunk 3]
```diff
@@ -1,36 +1,265 @@
...
+        </Container>
+    )
+}
+
+const validation = yup.object().shape({
+    firstName: yup.string().required('Required'),
+    lastName: yup.string().required('Required'),
+    email: yup.string().required('Required'),
+    password: yup.string().required('Required'),
+    terms: yup.bool().oneOf([true], 'Field must be checked'),
```
Comment (marckraw) [inline frontend/src/Landing.js]: Is i18n already set up on frontend ? I think i did that on the beggining but will double check, we can start using it right away i think, and probably easy translate error messages.

File: frontend/src/Landing.js [hunk 4]
```diff
@@ -1,36 +1,265 @@
...
+    }),
+    validationSchema: validation,
+    handleSubmit: async (values, bag) => {
+        bag.setStatus({ loading: true })
+        try {
+            await http.post('/auth/register', values)
+            bag.resetForm()
+            bag.setStatus({
+                loading: false,
+                successMessage: 'Your account has been successfully created.',
```
Comment (marckraw) [inline frontend/src/Landing.js]: Same thing as with error messages - i think we can start translating it from the beggining of the project, there is not that much hassle with it i think :)

File: frontend/src/Landing.js [hunk 5]
```diff
@@ -1,36 +1,265 @@
...
+                        </Grid>
+                        <Grid item xs={12}>
+                            <FormControl error={!!getErrorMessage('terms')}>
+                                <FormControlLabel
+                                    control={
+                                        <Checkbox
+                                            value="conditions"
+                                            color="primary"
+                                            onChange={() =>
+                                                props.setFieldValue(
```
Comment (marckraw) [inline frontend/src/Landing.js]: I guess u can also destructure `setFieldValue` from props, as other thing was destructed, to have consistency ?

File: frontend/src/Landing.js [hunk 6]
```diff
@@ -1,36 +1,265 @@
...
+import Typography from '@material-ui/core/Typography'
+import { makeStyles } from '@material-ui/core/styles'
+import Container from '@material-ui/core/Container'
+import SnackbarContent from '@material-ui/core/SnackbarContent'
+import IconButton from '@material-ui/core/IconButton'
+import CloseIcon from '@material-ui/icons/Close'
+import FormControl from '@material-ui/core/FormControl'
+import FormHelperText from '@material-ui/core/FormHelperText'
+import CircularProgress from '@material-ui/core/CircularProgress'
+import { green } from '@material-ui/core/colors'
```
Comment (marckraw) [inline frontend/src/Landing.js]: :o (nothing to add here, but it looks quite heavy on imports ;p)

File: backend/src/auth/auth.service.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+@Injectable()
+export class AuthService {
+    constructor(
+        private readonly usersService: UsersService,
+        private readonly jwtService: JwtService,
+        @InjectModel('User') private readonly userModel: Model<User>,
+    ) {}
+
+    async validateUser(username: string, pass: string): Promise<any> {
```
[the remark this answers is not in the document]
Comment (tomaszferens) [reply backend/src/auth/auth.service.ts]: this is the code from docs, will get rid of that "tutorial code", sorry, forgot to delete

File: backend/src/auth/auth.service.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        const user = await this.usersService.findOne(username)
+        if (user && user.password === pass) {
+            const { password, ...result } = user
+            return result
+        }
+        return null
+    }
+
+    async login(user: any) {
```
[the remark this answers is not in the document]
Comment (tomaszferens) [reply backend/src/auth/auth.service.ts]: as bove

File: backend/src/auth/jwt.strategy.ts (hunk 1 above)
Comment (tomaszferens) [reply backend/src/auth/jwt.strategy.ts]: as bove

File: frontend/src/Landing.js (hunk 6 above)
Comment (tomaszferens) [reply frontend/src/Landing.js]: yes, but I think it's worth it. Bundle size will be much smaller - which is what we want to support mobile first.

**Pull request merged**